packages/storage-plus/src/snapshot.rs

```diff
@@ -0,0 +1,371 @@
+    pub const fn new(namespaces: SnapshotNamespaces<'a>, strategy: Strategy) -> Self {
+        SnapshotMap {
+            primary: Map::new(namespaces.pk),
+            checkpoints: Map::new(namespaces.checkpoints),
+            changelog: Map::new(namespaces.changelog),
+            strategy,
+        }
+    }
+
+    pub fn add_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.checkpoints
+            .update::<_, StdError>(store, height.into(), |count| {
+                Ok(count.unwrap_or_default() + 1)
+            })?;
+        Ok(())
+    }
+
+    pub fn remove_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        let count = self
+            .checkpoints
+            .may_load(store, height.into())?
+            .unwrap_or_default();
+        if count <= 1 {
+            self.checkpoints.remove(store, height.into());
+            Ok(())
+        } else {
+            self.checkpoints.save(store, height.into(), &(count - 1))
+        }
+    }
+}
+
+impl<'a, K, T> SnapshotMap<'a, K, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a> + Prefixer<'a>,
+{
+    pub fn key(&self, k: K) -> Path<T> {
+        self.primary.key(k)
+    }
+
+    pub fn prefix(&self, p: K::Prefix) -> Prefix<T> {
+        self.primary.prefix(p)
+    }
+
+    /// should_checkpoint looks at the strategy and determines if we want to checkpoint
+    fn should_checkpoint(&self, store: &dyn Storage, k: &K) -> StdResult<bool> {
+        match self.strategy {
+            Strategy::EveryBlock => Ok(true),
+            Strategy::Never => Ok(false),
+            Strategy::Selected => {
+                // most recent checkpoint
+                let checkpoint = self
+                    .checkpoints
+                    .range(store, None, None, Order::Descending)
+                    .next()
+                    .transpose()?;
+                if let Some((height, _)) = checkpoint {
+                    // any changelog for the given key since then?
+                    let start = Bound::inclusive(U64Key::from(height));
+                    let first = self
+                        .changelog
+                        .prefix(k.clone())
+                        .range(store, Some(start), None, Order::Ascending)
+                        .next()
+                        .transpose()?;
+                    if first.is_none() {
+                        // there must be at least one open checkpoint and no changelog for the given address since then
+                        return Ok(true);
+                    }
+                }
+                // otherwise, we don't save this
+                Ok(false)
+            }
+        }
+    }
+
+    /// load old value and store changelog
+    fn write_change(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
+        let old = self.may_load(store, k.clone())?;
```

Comment:
```suggestion
        let old = self.primary.may_load(store, k.clone())?;
```
for clarity and consistency.